lib/utils/parse/ast.go

```diff
@@ -0,0 +1,487 @@
+			teleport.TraitExternalPrefix,
+		)
+	}
+}
+
+// buildEmailLocalExpr builds a EmailLocalExpr.
+func buildEmailLocalExpr(emailArg any) (Expr, error) {
+	// Validate first argument.
+	var email Expr
+	switch v := emailArg.(type) {
+	case string:
+		email = &StringLitExpr{value: v}
+	case Expr:
+		if v.Kind() == reflect.String {
+			email = v
+		}
+	}
+	if email == nil {
+		return nil, trace.BadParameter(
+			"found function %q with 1st argument that does not evaluate to a string",
+			EmailLocalFnName,
+		)
+	}
+	return &EmailLocalExpr{email: email}, nil
+}
+
+// buildRegexpReplaceExpr builds a RegexpReplaceExpr.
+func buildRegexpReplaceExpr(sourceArg, matchArg, replacementArg any) (Expr, error) {
+	// Validate first argument.
+	var source Expr
+	switch v := sourceArg.(type) {
+	case string:
+		source = &StringLitExpr{value: v}
+	case Expr:
+		if v.Kind() == reflect.String {
+			source = v
+		}
+	}
+	if source == nil {
+		return nil, trace.BadParameter(
+			"found function %q with 1st argument that does not evaluate to a string",
+			RegexpReplaceFnName,
+		)
+	}
+
+	// Validate second argument.
+	match, ok := matchArg.(string)
+	if !ok {
+		return nil, trace.BadParameter(
+			"found function %q with 2nd argument that is not a string",
+			RegexpReplaceFnName,
+		)
+	}
+	re, err := regexp.Compile(match)
+	if err != nil {
+		return nil, trace.BadParameter(
+			"failed to parse %q 2nd argument regexp %q: %v",
+			RegexpReplaceFnName,
+			match,
+			err,
+		)
+	}
+
+	// Validate third argument.
+	replacement, ok := replacementArg.(string)
+	if !ok {
+		return nil, trace.BadParameter(
+			"found function %q with 3rd argument that is not a string",
+			RegexpReplaceFnName,
+		)
+	}
+
+	return &RegexpReplaceExpr{source: source, re: re, replacement: replacement}, nil
+}
+
+// buildRegexpMatchExprFromLit builds a RegexpMatchExpr from a string literal.
+func buildRegexpMatchExprFromLit(raw string) (Expr, error) {
+	match, err := newRegexp(raw, true)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &RegexpMatchExpr{re: match}, nil
+}
+
+// buildRegexpMatchExpr builds a RegexpMatchExpr.
+func buildRegexpMatchExpr(matchArg any) (Expr, error) {
+	re, err := buildRegexpMatchFnExpr(RegexpMatchFnName, matchArg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &RegexpMatchExpr{re: re}, nil
+}
+
+// buildRegexpNotMatchExpr builds a RegexpNotMatchExpr.
+func buildRegexpNotMatchExpr(matchArg any) (Expr, error) {
+	re, err := buildRegexpMatchFnExpr(RegexpNotMatchFnName, matchArg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &RegexpNotMatchExpr{re: re}, nil
+}
+
+func buildRegexpMatchFnExpr(functionName string, matchArg any) (*regexp.Regexp, error) {
+	// Validate first argument.
+	// For now, only support a single match expression. In the future, we could
+	// consider handling variables and transforms by propagating user traits to
+	// the matching logic. For example
+	// `{{regexp.match(external.allowed_env_trait)}}`.
```

Comment:
this should be pretty easy to support now right, you would just have to compile the regexp during Evaluate? Do you plan to do that in another PR or are we intentionally limiting this to string literals for now?

Reply:
Yes, this is now very easy to support in the parser.

As you mentioned, we would just have to compile it during `Evaluate`. We could even optimize this and compile it before if we detected a constant expression (i.e. no vars) like `regexp.match("a")` or `regexp.match(email.local("vitor@goteleport.com"))`.

The part that it's missing is propagating user traits to all `Match` calls (e.g. [here](https://github.com/gravitational/teleport/blob/240739ecb5638d2e915b85b052ede272d7967aaa/lib/services/access_request.go#L772) and [here](https://github.com/gravitational/teleport/blob/240739ecb5638d2e915b85b052ede272d7967aaa/lib/services/role.go#L1855)).

I would be happy to do this in a follow up PR. This PR requires the argument to be a string literal just to keep the current behaviour.